### ISSUE TITLE
togo.webview can provide a jsbridge to call Python functions like PyQt5 webview

### ISSUE THREAD

**Issue by appleman4000** [beeware/toga#2268]
### What is the problem or limitation you are having?

PyQt5 WebView not work in iOS or Android so I hope toga can realize this function

### Describe the solution you'd like

this is pyqt5 solution

**index.html:**
```html
<!DOCTYPE html>
<html>
<head>
    <title>Webview Demo</title>
    <script type="text/javascript">
        function callPython() {
            alert(api.print('print from JavaScript!'));
        }
    </script>
</head>
<body>
    <button onclick="callPython()">Call Python Code print</button>
</body>
</html>
```

**webview.py:**
```python
import os

from PyQt5.QtCore import QUrl
from PyQt5.QtWebKit import QWebSettings
from PyQt5.QtWebKitWidgets import QWebView, QWebPage
from PyQt5.QtWidgets import QWidget

from posmesh.platform import API


class WebPage(QWebPage):
    def __init__(self, parent=None):
        super(WebPage, self).__init__(parent)
        self.api = API()
	self.mainFrame().javaScriptWindowObjectCleared.connect(self.populateJavaScriptWindowObject)
    def populateJavaScriptWindowObject(self):
        self.mainFrame().addToJavaScriptWindowObject('api', self.api)
		
class Webview(QWidget):
    def __init__(self, url="", parent=None):
        super(Webview, self).__init__(parent)
        self.view = QWebView(self)
        self.view.setPage(WebPage())
        self.view.settings().setAttribute(QWebSettings.LocalContentCanAccessRemoteUrls, True)
        self.view.load(QUrl.fromLocalFile(os.getcwd() + url))  
        self.view.show()

```

**api.py**
```python
from PyQt5.QtCore import pyqtSlot as Slot, QObject
class API(QObject):
    @Slot(str, result=str)
    def print(self, msg):
        print(msg)
        return 'printed'
```

### Describe alternatives you've considered

Haven't found a WebView jsbridge solution across iOS, Android, windows, and Linux yet using python.

**Comment by freakboy3742** [comment]
Agreed this would be useful mechanism to have in Toga's WebView API. If nothing else, it would be necessary for a full "Electron"-style app shell replacement, so that web code can invoke native system functions.

The implementation is going to be very platform specific. From a quick poke around, here are some pointers on the APIs that will be required:

* iOS/Cocoa (both using WKWebview) [This is an example of usage](https://stackoverflow.com/a/61182119)
* Android: [This is an example of usage](https://stackoverflow.com/a/65814170)
* GTK: [this will be the API to use](https://webkitgtk.org/reference/webkit2gtk/stable/method.UserContentManager.register_script_message_handler.html); the usage will be very similar to iOS, as they're both WebKit based
* Windows: [This is a tutorial on how to use the API](https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/hostobject?tabs=win32); interacting with Python.net could be complicated as it's a COM interface.

Ultimately, we need to be able to have a `webview.register_handler("name", handler_method)` API of some kind; however, in the meantime, you can use the platform specific APIs by dropping to the native interface. On any user-space toga.WebView instance, `webview._impl.native` will be the platform native WKWebKit/GTK.WebView2/etc instance; it should be possible to use the example code provided above on this native object.

**Comment by proneon267** [comment]
FWIW, this can be currently done with: https://github.com/python-eel/Eel. It can be combined with toga's view for calling python from js.

**Comment by freakboy3742** [comment]
@proneon267 Are you suggesting that as something someone could drop in to a Toga app? Or an example of someone else implementing the equivalent functionality (using the APIs that I highlighted in my previous post)?

**Comment by proneon267** [comment]
Yes, someone can just use it in toga as is. I had tested the (toga+python-eel combo) and it had worked properly on desktop and mobile platform like Android.

**Comment by freakboy3742** [comment]
That... seems very unusual to me. Looking at the code, it only supports Chrome and Edge - so it can't support Toga's macOS interface - and it doesn't include any code differentiating Android from Linux; and it doesn't include any hooks into the native platform webview APIs. Can you provide an example of how it is used?

**Comment by proneon267** [comment]
Sure. I'll find the old project I was working on and get back to you.

**Comment by proneon267** [comment]
@freakboy3742 I have gutted out the complex logic in my old project and replaced it with a simpler logic. Here is the example briefcase project: https://github.com/proneon267/toga-python-eel. 

Particularly, take a look at: 
`app.py`: https://github.com/proneon267/toga-python-eel/blob/main/togapythoneel/src/togapythoneel/app.py and `script.js`: https://github.com/proneon267/toga-python-eel/blob/main/togapythoneel/src/togapythoneel/web/script.js

I have tested the example on Windows, Android and MacOS. Everything seems to be working properly. 

I also tried to run it on iOS but got an error because, gevent is not in the Anaconda iOS repository:https://anaconda.org/beeware/repo. So, it seems it is just a matter of building the binary wheels for gevent.

**Comment by freakboy3742** [comment]
Ah - you're not using Eel's core functionality. 

What you'e done here is use Eel to start a web server; when the web page hits a URL, it hits the web server, which is in Python space, which allows you to run Python code. Eel might make it marginally easier to start a web server... but you could do that with *any* web server (Django, Flask, or even a `http.server` subclass). 

That's fundamentally different to what this ticket is proposing - the approach suggested by this ticket wouldn't require a standalone webserver. It would integrate directly with the webview, so that invoking Javascript calls back into the same process that is hosting the web view.

**Comment by proneon267** [comment]
Oh - in that case, it will need to be implemented in toga 😅

**Comment by freakboy3742** [comment]
An implementation note: If this is resolved for GTK or Winforms, the MapView widget introduced by #2379 will be able to implement the `on_select` handler.

**Comment by freakboy3742** [comment]
Here's a worked example of a webview that will work with Toga 0.4.8 on Cocoa; it will likely work on iOS as well.

It's a fairly vanilla webview example - a Toga app that only contains a webview, loaded with static HTML content. That HTML content displays a button; when that button is pressed, it passes a message.

On the Python side, the WebView is constructed with a MessageHandler that can catch that message (in `userContentController_didReceiveScriptMessage_()`). In this example, it responds by showing a dialog if the message matches, but it could do anything else you want - when this gets in to main, it should probably be invoking an `on_javascript_message()` hander (or similar).

It requires patching the cocoa WebView's `create()` method; this is needed because the WebView configuration needs to be set at time of construction. The definition of the configuration object and the message handler should be migrated into the cocoa WebView implementation.

```
import toga
from toga.style import Pack

from toga_cocoa.widgets import webview
from rubicon.objc import objc_method, objc_property, ObjCClass, NSObject, NSMakeRect

WKWebViewConfiguration = ObjCClass("WKWebViewConfiguration")
WKUserContentController = ObjCClass("WKUserContentController")


class TogaScriptMessageHandler(NSObject):
    interface = objc_property(object, weak=True)
    impl = objc_property(object, weak=True)

    @objc_method
    def userContentController_didReceiveScriptMessage_(
        self, userContentController, message
    ) -> None:
        print("Script message: ", message.body)
        if message.body == "open_dialog":
            self.interface.app.loop.create_task(
                self.interface.window.dialog(
                    toga.InfoDialog("Hi there!", "This is a native dialog")
                )
            )
        else:
            print("UNKNOWN MESSAGE")


orig_create = webview.WebView.create


def monkey_create(self):
    # Content that should probably be in cocoa's WebView
    configuration = WKWebViewConfiguration.alloc().init()
    configuration.userContentController = WKUserContentController.alloc().init()

    message_handler = TogaScriptMessageHandler.alloc().init()
    message_handler.impl = self
    message_handler.interface = self.interface

    configuration.userContentController.addScriptMessageHandler(
        message_handler,
        name="toga_handler",
    )

    # An update to the default instantiation of Webview
    self.native = webview.TogaWebView.alloc().initWithFrame(
        # A dummy size that will be immediately updated by constraints
        NSMakeRect(0, 0, 0, 0),
        configuration=configuration,
    )

    # The rest of the normal Cocoa webview creation.
    self.native.interface = self.interface
    self.native.impl = self

    self.native.inspectable = True
    self.native.navigationDelegate = self.native

    self.loaded_future = None

    self.add_constraints()


webview.WebView.create = monkey_create


class ExampleWebView(toga.App):

    def startup(self):
        self.main_window = toga.MainWindow()

        self.webview = toga.WebView(style=Pack(flex=1))

        self.webview.set_content(
            None,
            """<!DOCTYPE html>
<html>
    <body>
        <button onclick="onClick()">Click me to open a native dialog</button>
        <script>
            function onClick() {
                window.webkit.messageHandlers.toga_handler.postMessage("open_dialog");
            }
        </script>
    </body>
</html>
""",
        )

        self.main_window.content = self.webview
        self.main_window.show()


def main():
    return ExampleWebView("Toga WebView Demo", "org.beeware.toga.examples.webview")


if __name__ == "__main__":
    app = main()
    app.main_loop()
```

**Comment by destroyer66** [comment]
> [@freakboy3742](https://github.com/freakboy3742) I have gutted out the complex logic in my old project and replaced it with a simpler logic. Here is the example briefcase project: https://github.com/proneon267/toga-python-eel.
> 
> Particularly, take a look at: `app.py`: https://github.com/proneon267/toga-python-eel/blob/main/togapythoneel/src/togapythoneel/app.py and `script.js`: https://github.com/proneon267/toga-python-eel/blob/main/togapythoneel/src/togapythoneel/web/script.js
> 
> I have tested the example on Windows, Android and MacOS. Everything seems to be working properly.
> 
> I also tried to run it on iOS but got an error because, gevent is not in the Anaconda iOS repository:https://anaconda.org/beeware/repo. So, it seems it is just a matter of building the binary wheels for gevent.
> 
> Looks like we are closer to cross-platform Electron-Style apps in Python :)



引用你这个测试了一下，在Windows11下没有发现任何问题，但是在编译Android APK的时候，就遇到问题了，提示以下错误信息：

building 'greenlet._greenlet' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/

可以很明显地看出这个eel依赖于Windows，因为它写了需要Microsoft Visual C++ 14.0，有没有更好的解决方案呢？
